crawl-ref/source/menu.cc

```diff
@@ -1415,63 +1606,29 @@ void Menu::draw_index_item(int index, const MenuEntry *me) const
     if (crawl_state.doing_prev_cmd_again)
         return;
 
-    draw_stock_item(index, me);
-}
-
-void Menu::draw_stock_item(int index, const MenuEntry *me) const
-{
     mdisplay->draw_stock_item(index, me);
 }
 
 bool Menu::page_down()
 {
-    int old_first = first_entry;
-
-    if ((int) items.size() > first_entry + pagesize)
-    {
-        first_entry += pagesize;
-        //if (first_entry + pagesize > (int) items.size())
-        //    first_entry = items.size() - pagesize;
-
-        if (old_first != first_entry)
-            return true;
-    }
-    return false;
+    return mdisplay->scroll_page(1);
 }
 
 bool Menu::page_up()
 {
-    int old_first = first_entry;
-
-    if (first_entry > 0)
-    {
-        if ((first_entry -= pagesize) < 0)
-            first_entry = 0;
-
-        if (old_first != first_entry)
-            return true;
-    }
-    return false;
+    return mdisplay->scroll_page(-1);
 }
 
 bool Menu::line_down()
 {
-    if (first_entry + pagesize < (int) items.size())
-    {
-        ++first_entry;
-        return true;
-    }
-    return false;
+    MenuDisplayText *text_display = dynamic_cast<MenuDisplayText*>(mdisplay);
```

Comment:
There's a bunch of static/dynamic casting of mdisplay to `MenuDisplayText *` that make me wonder if the encapsulation is quite right yet.

I'm also not sure I understand the motivation for removing line-by-line scrolling from MenuDisplayTiles, which would clean up this particular cast. For me at least, this is the default way I try to interact with crawl menus so I think it is a bit surprising from the user's perspective that it won't work.  I guess the problem is that hotkey assignment is tricky, but I would still rather keep some version of the current behavior.